nexus/src/app/sagas/instance_update/start.rs

```diff
@@ -189,30 +191,102 @@ async fn siu_fetch_state_and_start_real_saga(
+            .sagas
+            .saga_prepare(dag)
             .await
-        {
-            warn!(
-                log,
-                "instance update: real update saga failed (which *could* \
-                 mean nothing...)";
-                "instance_id" => %instance_id,
-                "error" => %error,
-            );
-            // If the real saga failed, kick the background task. If the real
-            // saga failed because this action was executed twice and the second
-            // child saga couldn't lock the instance, that's fine, because the
-            // background task will only start new sagas for instances whose DB
-            // state actually *needs* an update.
-            osagactx.nexus().background_tasks.task_instance_updater.activate();
+            // Similarly, if we can't prepare the child saga, we need to unwind
+            // and release the lock.
+            .map_err(|e| {
+                nexus.background_tasks.task_instance_updater.activate();
+                ActionError::action_failed(e)
+            })?
+            .start()
+            .await
+            // And, if we can't start it, we need to unwind.
+            .map_err(|e| {
+                nexus.background_tasks.task_instance_updater.activate();
+                ActionError::action_failed(e)
+            })?
+            .wait_until_stopped()
+            .await
+            .into_raw_result();
+        match child_result.kind {
+            Ok(_) => {
+                debug!(
+                    log,
+                    "instance update: child saga completed successfully";
+                    "instance_id" => %instance_id,
+                    "child_saga_id" => %child_result.saga_id,
+                )
+            }
+            // Check if the child saga failed to inherit the updater lock from
+            // this saga.
+            Err(SagaResultErr {
+                error_node_name,
+                error_source: ActionError::ActionFailed { source_error },
+                ..
+            }) if error_node_name.as_ref() == super::INSTANCE_LOCK => {
+                if let Ok(instance::UpdaterLockError::AlreadyLocked) =
+                    serde_json::from_value(source_error)
+                {
+                    // If inheriting the lock failed because the lock was held by another
+                    // saga. If this is the case, that's fine: this action must have
+                    // executed more than once, and created multiple child sagas. No big deal.
+                    return Ok(());
+                } else {
+                    // Otherwise, the child saga could not inherit the lock for
+                    // some other reason. That means we MUST unwind to ensure
+                    // the lock is released.
+                    return Err(ActionError::action_failed(
+                        "child saga failed to inherit lock".to_string(),
+                    ));
+                }
```

Comment:
There are a few edge cases here that I think the system will handle correctly, but I want to work through them here to see if you agree (and to see if you have further comments about them).

## Case 1: Parent spawns multiple child sagas; one succeeds, then one fails due to a DB query failure

1. Parent saga creates child C1
2. Parent saga crashes and replays, creating child C2
3. C1 inherits the lock successfully
4. C2's call to CRDB fails

In this case, P will try to unwind and call `instance_updater_unlock`; this will return `Err` because the lock IDs no longer match. However, `siu_lock_instance_undo` swallows this error, so the unwinding parent won't get stuck. (**N.B.** If this is right it might be good to add a comment to `siu_lock_instance_undo` explaining why it is sometimes expected to get a lock ID mismatch here.)

## Case 2: Parent spawns multiple child sagas; one fails due to a DB query failure, then the other succeeds

1. Parent saga creates child C1
2. Parent saga crashes and replays, creating child C2
3. C2 fails and returns `UpdaterLockError::Query`, causing P to unwind
7. C1 tries to inherit the lock

Here, C1 will fail to inherit because its predecessor lock is gone, but that's OK--the instance won't be updated, but the lock isn't abandoned and the instance can be updated again.

## Case 3: Parent spawns multiple child sagas; the abandoned saga fails but its sibling succeeds

1. Parent saga creates child C1
2. Parent saga crashes and replays, creating child C2
3. C1 fails to inherit the lock

This can happen either because

a. C2 inherited the lock first and C1 returned `AlreadyLocked`: this is fine, C2 owns the lock now
b. C1 failed to query: this information is lost, but that's fine; C2's result will determine what the parent does, as though C1 had never been launched at all

## Case 4: Instance gets deleted before a child can try to inherit the lock

1. P spawns C1 and C2
2. C1 inherits and drops the lock
3. Instance gets deleted
4. C2 gets a query error because the instance is gone

This dovetails with case 1: P will try to unwind, and `instance_updater_unlock` will return an `Err`, except that its reason for doing so will be slightly different (instance ID not found rather than a lock ID mismatch). Either way P will swallow the error.

---

The general observation here is that, as long as the parent stops replaying this node, there will eventually be a "last" child saga that can do one of the following things:

a. succeed,
b. return `UpdaterLockError::AlreadyLocked`, which implies that someone else either is or was previously responsible for dropping the lock, or
c. return `UpdaterLockError::Query`, which causes the parent to unwind, which is acceptable even if some other (orphaned) child saga managed to inherit the lock in the meantime

And the key bit is that unwinding the parent saga even when someone else previously inherited is OK--the parent won't undercut the inheritor in that case because of the filters in `instance_updater_unlock`.

Reply:
These are also the cases I had in mind when I was writing this code, at least.

Honestly, trying to get this right _did_ kind of make me wonder whether attempting to prevent duplicate child sagas from running was even worth it. We _could_ just allow multiple children with the same `UpdateRequired` to proceed freely, and rely on saga action idempotency to assume that it's fine for there to be two (or more) of 'em...my initial motivation was that it was a waste of resources to serialize the same saga twice. But, we _could_ probably kill the inherit-lock operation and everything would still be _correct_.

Reply:
OTTOH, allowing duplicate update sagas is only okay because the actual update of the instance record is conditional on the state generation not having changed, to prevent ABA problems. Otherwise, there could be a failure mode with two child update sagas in which one update saga writes the new state and unlocks the instance, somebody else does something to the now-unlocked instance, and then the other update saga wakes up and puts the instance state _back_ to the way it was before. But, we already have generation numbers for that.